Nginx :  ngx_configure_listening_sockets and hijack `recvmsg`

firesidefox opened the pull request:
1. Nginx based on fstack delays setting up server on fstack until ngx_worker_process_init. ngx_configure_listening_sockets should as well.
2. FStack does not support IP_PKTINFO

Fix #157 .

Commit:
Nginx :  ngx_configure_listening_sockets
1. Nginx based on fstack delays setting up server on fstack until ngx_worker_process_init. ngx_configure_listening_sockets should as well.
2. FStack does not support IP_PKTINFO

Commit:
Nginx: hijack `recvmsg`.
Since Nginx calls `recvmsg` in stream with udp, we must
hijack this function, so that the network IO can pass through
f-stack.

whl739 commented on app/nginx-1.11.10/src/core/ngx_connection.c:
```diff
@@ -13,6 +13,33 @@
+                            ngx_process);                                     \
+        exit(1);                                                              \
+    }                                                                   
 #endif
```
Is it better to use a function than macro here?

firesidefox replied:
On this, macro and function are same, encapsulates the public operation to carry out code reuse. 
Macro maybe have a better performance compared with (inline) function.  
What do you think? Have I missed something?

whl739 replied:
> On this, macro and function are same, encapsulates the public operation to carry out code reuse.
Macro maybe have a better performance compared with (inline) function.

You are right, they are same, and macro has a better performance, but it has no effect, it's only called during initializing.
I just thought that the code is a little more here.

whl739 commented on app/nginx-1.11.10/src/core/ngx_connection.c:
```diff
@@ -411,31 +438,7 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
-                                  ngx_process);
-                exit(1);
-            }
+            ngx_ff_skip_listening_socket(ls);
```
It's better to pass `ls[i]` to `ngx_ff_skip_listening_socket`.

Commit:
Nginx : replace macro with inline function to implement ```ngx_ff_skip_listening_socket```

Commit:
API : fix bug of ff_sendmsg and ff_recvmsg
freebsd and linux have a different "struct sockaddr".
In ff_recvmsg and ff_recvmsg, ```msg->msg_name```,which can refer to address, can be expected to be converted .

whl739 commented on lib/ff_syscall_wrapper.c:
```diff
@@ -610,6 +610,11 @@ ff_sendmsg(int s, const struct msghdr *msg, int flags)
     int rc;
 
+    if (msg->msg_name != NULL) {
+        linux2freebsd_sockaddr(msg->msg_name,
```
`msg`'s value has been modified, it's against with the const qualifier.
Better like this:
```
void *oldsa = msg->msg_name;
struct sockaddr sa;
linux2freebsd_sockaddr(msg->msg_name, sizeof(struct linux_sockaddr), &sa);
msg->msg_name = &sa;
sendit..
msg->msg_name = oldsa;
```
or
```
struct msghdr m;
memcpy(&m, msg, sizeof(struct msghdr));
if (m.msg_name != NULL) {
    linux2freebsd_sockaddr(m->msg_name, sizeof(struct linux_sockaddr), m->msg_name);
}
```

Commit:
API : ff_sendmsg, use transient variables to avoid polluting user's data.
